src/cycles/eulerian.jl

```diff
@@ -0,0 +1,315 @@
+            edge_count[next_vertex] -= 1
+            current_vertex = next_vertex
+        else
+            push!(circuit, current_vertex)
+            current_vertex = current_path[end]
+            pop!(current_path)
+        end
+    end
+
+    return reverse!(circuit)
+end
+
+"""
+    eulerian_trail(g)
+
+Return array for vertices representing Eulerian trail if graph `g` contains [Eulerian trail](https://en.wikipedia.org/wiki/Eulerian_path) otherwise throws error.
+
+### Implementation Notes
+Implemented Hierholzer's algorithm to find Eulerian trail.
+The idea is to keep following unused edges and removing them until we get stuck.
+Once we get stuck, we back-track to the nearest vertex in our current path that has unused edges, and we repeat the process until all the edges have been used.
+For undirected graph, made a copy of adjacency list and deleted edges from it while iterating.
+For directed graph, since each edge is in one vertex's adjacency, it can be virtually removed it by decrementing the edge_count.
+For Eulerian trail we start tracking the path from vertices with odd degree in case of undirected graph and for directed graph from a vertex having outdegree - indegree = 1, if exists.
+Since degree() accounts only once of self loop in directed graph, incremented degree by 1 for the algorithm to work in undirected graphs.
+"""
+function eulerian_trail end
+@traitfn function eulerian_trail(
+    graph::AG::(IsDirected)
+    ) where {T, AG <: AbstractGraph{T}}
+    
+    has_eulerian_trail(graph) || throw(ArgumentError("Graph do not have Eulerian trail."))
+    trail = Vector{T}()
+    iszero(nv(graph)) && return trail
+    iszero(ne(graph)) && return [vertices(graph)[1]]
+
+    edge_count = collect(outdegree(graph))
+    current_vertex = 1
+    for v in vertices(graph)
+        if outdegree(graph, v) - indegree(graph, v) == 1
+            current_vertex = v
+            break
+        end
+    end
+    current_path = Vector{T}([current_vertex])
+
+    while !isempty(current_path)
+        if !iszero(edge_count[current_vertex])
+            push!(current_path, current_vertex)
+            next_vertex = neighbors(graph, current_vertex)[edge_count[current_vertex]]
+            edge_count[current_vertex] -= 1
+            current_vertex = next_vertex
+        else
+            push!(trail, current_vertex)
+            current_vertex = current_path[end]
+            pop!(current_path)
+        end
+    end
+    
+    return reverse!(trail)
+end
+
+@traitfn function eulerian_trail(
+    graph::AG::(!IsDirected)
+    ) where {T, AG <: AbstractGraph{T}}
+
+    has_eulerian_trail(graph) || throw(ArgumentError("Graph do not have Eulerian trail."))
+    trail = Vector{T}()
+    iszero(nv(graph)) && return trail
+    iszero(ne(graph)) && return [vertices(graph)[1]]
+
+    current_vertex = 1
+    edge_count = collect(outdegree(graph))
+    for v in vertices(graph)
+        has_edge(graph, v, v) && (edge_count[v] += 1)
```

Comment:
I don't think we should use the shortcut with `&&` with mutating statements such as `edge_count[v] += 1`, as this might easily be overlooked. Can you replace that by an if statement?

Reply:
sure